br/pkg/lightning/backend/local/local.go

```diff
@@ -1541,53 +1604,89 @@ func (local *local) writeAndIngestByRanges(ctx context.Context, engine *Engine,
 	var allErr error
 	var wg sync.WaitGroup
 	metErr := atomic.NewBool(false)
+	storeId2Range := make(map[uint64][]Range)
+	storeId2RangeIndex := make(map[uint64]int)
 
 	for _, r := range ranges {
 		startKey := r.start
 		endKey := r.end
-		w := local.rangeConcurrency.Apply()
-		// if meet error here, skip try more here to allow fail fast.
-		if metErr.Load() {
-			local.rangeConcurrency.Recycle(w)
-			break
+		regions, err := local.scanRegion(ctx, engine, startKey, endKey, regionSplitSize, regionSplitKeys)
+		if err != nil {
+			log.FromContext(ctx).Warn("scaen regino failed", zap.Error(err))
+			return err
 		}
-		wg.Add(1)
-		go func(w *worker.Worker) {
-			defer func() {
-				local.rangeConcurrency.Recycle(w)
-				wg.Done()
-			}()
-			var err error
-			// max retry backoff time: 2+4+8+16+30*26=810s
-			backOffTime := time.Second
-			for i := 0; i < maxWriteAndIngestRetryTimes; i++ {
-				err = local.writeAndIngestByRange(ctx, engine, startKey, endKey, regionSplitSize, regionSplitKeys)
-				if err == nil || common.IsContextCanceledError(err) {
-					return
-				}
-				if !local.isRetryableImportTiKVError(err) {
-					break
-				}
-				log.FromContext(ctx).Warn("write and ingest by range failed",
-					zap.Int("retry time", i+1), log.ShortError(err))
-				backOffTime *= 2
-				if backOffTime > maxRetryBackoffTime {
-					backOffTime = maxRetryBackoffTime
-				}
-				select {
-				case <-time.After(backOffTime):
-				case <-ctx.Done():
-					return
-				}
+		// only take to leader's store id fo the first region
+		if len(regions) > 0 {
+			if val, ok := storeId2Range[regions[0].Leader.StoreId]; ok {
+				storeId2Range[regions[0].Leader.StoreId] = append(val, r)
+			} else {
+				storeId2Range[regions[0].Leader.StoreId] = []Range{r}
 			}
+		}
+	}
 
-			allErrLock.Lock()
-			allErr = multierr.Append(allErr, err)
-			allErrLock.Unlock()
-			if err != nil {
-				metErr.Store(true)
+	for i := range storeId2Range {
+		storeId2RangeIndex[i] = len(storeId2Range[i])
+	}
+
+	finished := 0
+outerLoop:
+	for finished < len(storeId2RangeIndex) {
+		for storeId := range storeId2RangeIndex {
+			if storeId2RangeIndex[storeId] > 0 {
+				dataRange := storeId2Range[storeId][storeId2RangeIndex[storeId]-1]
+				startKey := dataRange.start
+				endKey := dataRange.end
+				w := local.rangeConcurrency.Apply()
+				// if meet error here, skip try more here to allow fail fast.
+				if metErr.Load() {
+					local.rangeConcurrency.Recycle(w)
+					break outerLoop
+				}
+				wg.Add(1)
+				go func(w *worker.Worker) {
+					defer func() {
+						local.rangeConcurrency.Recycle(w)
+						wg.Done()
+					}()
+					var err error
+					// max retry backoff time: 2+4+8+16+30*26=810s
+					backOffTime := time.Second
+					for i := 0; i < maxWriteAndIngestRetryTimes; i++ {
+						err = local.writeAndIngestByRange(ctx, engine, startKey, endKey, regionSplitSize, regionSplitKeys)
```

Comment:
there's also a PaginateScanRegion inside writeAndIngestByRange. Because we just get the region of every key range in the new scanRegion, I think we can skip the PaginateScanRegion inside writeAndIngestByRange.

It may be a bit hard for you to change the related codes, and we expect the code changes should be made on master branch first. Maybe we can implement your idea on master branch in out separate PR. We will discuss soon.

Reply:
cool, if you could help to implement on master branch, I could cherry pick it to our 6.5 branch